### DISCUSSION TITLE
What would be the impact of Breaking Change for tree-sitter-lua on `neovim` core, `nvim-treesitter` and plugins?

### DISCUSSION THREAD

**Discussion by MunifTanjim** [nvim-treesitter/nvim-treesitter#4706] [General]
If we want to add node for escaped string in `tree-sitter-lua` (https://github.com/MunifTanjim/tree-sitter-lua/issues/9), what would be the impact of that breaking change?

It would affect:
- `neovim` core
- `nvim-treesitter`
- plugins

Would it be worth it? And what would be the recommended approach if it's worth it?

cc: @amaanq @clason @theHamsta

**Comment by amaanq** [comment]
Doesn't really have any impact, all it'd take is adding the escape_sequence rule as a `@string.escape` capture

**Comment by clason** [comment]
> breaking change?

Would that actually be breaking? I mean, would the old queries no longer work?

> Would it be worth it? 

If it's useful, sure. (I couldn't say, personally ;))

> And what would be the recommended approach if it's worth it?

If it's breaking, make a new release (with updated upstream queries). We'll just bump the requirement in Neovim (and probably backport to 0.9.1 to avoid diverging needlessly from nvim-treesitter). To be clear: Things are set up so that we can (mostly) handle incompatible parser versions in Neovim and nvim-treesitter.

Having releases makes this easier for us.

**Reply by amaanq** [reply]
I submitted the PR to add that functionality so I'll drop my 2 cents:

> Would that actually be breaking? I mean, would the old queries no longer work?

No

> If it's useful, sure. (I couldn't say, personally ;))

I'd say it is :)

> If it's breaking 

No since the current queries *would* work, but the only "issue" I could think of if the parser were to be updated w/o its queries is that escape sequences wouldn't be highlighted at all, but regular string content bits would be fine (and again, it's adding one capture)

**Reply by clason** [reply]
But thank you for asking! You're the first person to be that considerate ❤️ 

Even though this specific instance is harmless, we could take this opportunity to have a general discussion about breaking changes in the tree-sitter ecosystem. I would like to push more parsers into following proper semver to make maintenance easier for everyone. (E.g., point release for changes to the grammar that doesn't affect parsers, minor releases for changes that _add_ nodes so old queries continue to work, major releases for changes that break existing queries because they remove/rename nodes. For the mythical nvim-treesitter 2.0, we could restrict the top tier to parsers that make releases following semver and bump everything else to a more experimental tier.)
